docs/design_notes/module_lifecycle.md

```diff
@@ -0,0 +1,289 @@
+
+Thus, what remains is to actually understand what `.apply(...)` does. All the interesting stuff in a Module happens inside the `apply` method. We will refer to the various stages before and during `apply` as the “Module lifecycle”. We can summarize this lifecycle in a number of steps:
+
+
+1. We construct `mlp = MLP(5, 3)`, such that `mlp.hidden_size=5` and `mlp_out_size=3`
+2. Then, `mlp.apply`:
+    1. Makes a clone of `mlp`, let’s call it `mlp_copy`
+    2. Calls `mlp_copy.setup()`.
+    3. Returns the output of `mlp_copy.__call__()` and optionally the variable collections that were specified as mutable using `mutable=` kwarg.
+
+
+You might be surprised that the lifecycle includes cloning the Module instance. This will be explained later in the **Top-level Modules** section.
+
+### Variables
+
+The word “variable” is ubiquitous in programming and math. However, it’s important to have a good understanding of what that means in the context of JAX and Flax. Inside Flax modules, variables act like you expect from Python. They are initialized once, read, and perhaps even updated every so often. However, JAX has no concept of variables. Instead, values are stored in arrays similar to NumPy arrays  -- with one important difference: they are immutable.
+
+`init` and `apply` return the variables as a dictionary of ordinary JAX arrays. They are immutable and therefore really just a snapshot of the variables. Next time we call apply we can feed back the snapshot so the variables are again in the same state as where we finished on the last call to apply.
+
+
+> **NOTE** \
+Module fields are declared using the `field_name: TypeHint` syntax (same as dataclasses). Without a type hint, an attribute is considered a static property of the class. In case you cannot specify the type you can use `typing.Any` as a wildcard type.
+
+
+## Compact Modules
+
+Linen provides an alternative API for defining modules more compactly. This is especially useful for the common case where the Module consists of only one method that uses parameters and/or sub-modules. Using the compact API the MLP can be rewritten as follows:
+
+
+```python
+class CompactMLP(nn.Module):
+  hidden_size: int
+  out_size: int
+
+  @nn.compact
+  def __call__(self, x):
+    a = nn.Dense(self.hidden_size)(x)
+    h = nn.relu(a)
+    return nn.Dense(self.out_size)(h)
+
+```
+
+
+A compact `Module` is similar in spirit to a function. It offers a concise notation and restricts external interaction to the inputs and return values of the function. In this case the concise notation might make it easier for others to understand what the Module is doing. There is no need to jump back and forth between the setup and call function to understand what the submodules are doing. Instead, simply reading the call method from top to bottom once should give a clear picture of what the Module is doing. This can make a significant difference if you are implementing complex Modules with many hyper parameters.
```

Comment:
```suggestion
A `compact` `Module` is similar in spirit to a function. It offers a concise notation and restricts external interaction to the inputs and return values of the function. In this case the concise notation might make it easier for others to understand what the Module is doing. There is no need to jump back and forth between the setup and call function to understand what the submodules are doing. Instead, simply reading the call method from top to bottom once should give a clear picture of what the Module is doing. This can make a significant difference if you are implementing complex Modules with many hyper parameters.
```

Note that "compact", like "Module", are common English words, so using backticks here and there may help users understand the context better.